### ISSUE TITLE
Float16 tensors converted back to Float32 when passed to GPU

### ISSUE THREAD

**Issue by ThomasLecat** [MIC-DKFZ/medicaldetectiontoolkit#28]
Hi Paul,

I noticed that in the data loaders of the LIDC and toy experiments, the data was loaded as float16 tensors.

However, in all models, the data is allocated to a GPU, it seems to be converted back to Float32 as `.float()` always precede `.cuda()`. For example:

Retina U-Net
https://github.com/pfjaeger/medicaldetectiontoolkit/blob/2b343ba974a6936d10bc7ba2d571db28f0760df5/models/retina_unet.py#L398

Mask RCNN
https://github.com/pfjaeger/medicaldetectiontoolkit/blob/2b343ba974a6936d10bc7ba2d571db28f0760df5/models/mrcnn.py#L864

UFRCNN
https://github.com/pfjaeger/medicaldetectiontoolkit/blob/2b343ba974a6936d10bc7ba2d571db28f0760df5/models/ufrcnn.py#L822

Strange thing is, when training with Mask RCNN, I can double the batch_size before triggering a CUDA: out-of-memory error when the data is loaded as Float16 compared to Float32. 

I didn't test this for the other models, but do you have an idea about:
1. does the models use Float16 tensors in some places I couldn't spot?
2. do you get the the same behaviour regarding batch_size and CUDA OOM errors when switching from Float16 to Float32?
3. If the answer to 1 is no, what would cause the above mentioned behaviour? 

Thanks!

**Comment by pfjaeger** [comment]
Hi Thomas,
sorry for the extremely delayed response. I was not able to reproduce the described behaviour, so changing line 305 in lidc dataloader to np.float32 results in the same memory usage than np.float16 (I tested with 2D & 3D Mask R-CNN). I agree that it does not make sense to load data as float16 and I changed that in the latest commit. 

However, NVIDIA (and I heard also pytorch but with more flaws) provides a nice tool for float16 computations. Our tests show significant time and memory savings without performance loss!

See https://github.com/NVIDIA/apex/tree/master/apex/amp

This works only for 2D convolutions however, which is why I did not include it here at this point.
